### ISSUE TITLE
Enum consistency

### ISSUE THREAD

**Issue by DomHenderson** [lambdaclass/cairo_native#431]
We've been running into segaults when creating aot or jit native executor's for a fully featured openzeppelin erc20, that seem to be caused by the inclusion of code that emits the OwnableComponent's Event (summarised below). I'm working on a fix atm but would appreciate some input. Think I've tracked it down to inconsistencies between types/enum.rs>build and libfuncs/enum.rs>build_init and build_match. To confirm, is the model in types/enum.rs>build which has 4 different options: 1 that's just an array, and 3 that are structs, the definitive model to use? I've been working on making sure the rest of the codebase respects it, and have fixed some bugs in doing so, but I don't have the most solid grasp on the padding rules yet

```
#[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        #[flat]
        ERC20Event: ERC20Component::Event,
        #[flat]
        PausableEvent: PausableComponent::Event,
        #[flat]
        OwnableEvent: OwnableComponent::Event,
        #[flat]
        UpgradeableEvent: UpgradeableComponent::Event,
    }

#[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        OwnershipTransferred: OwnershipTransferred
    }

#[derive(Drop, starknet::Event)]
    struct OwnershipTransferred {
        previous_owner: ContractAddress,
        new_owner: ContractAddress,
    }
```

**Comment by azteca1998** [comment]
When we implemented AOT we realized that the previous enum representation was an absolute nightmare to use with anything other than `invoke_packed`. Basically, when invoking the function using the C interface, everything is "flattened".

For example, if we had a struct like this `(u32, u32, u32)` we needed to pass the three `u32` as separate arguments, as if the function was `fn(u32, u32, u32)`. I think it was a bit more complicated than that, but this is enough to explain the change. If we had an enum then we needed to pass a lot of useless arguments (every byte of padding is an argument) in the best of cases, while having to juggle with bytes of data just to call the function at worst (overlapping incompatible variants).

Because of that, we changed the enums so that:
  - Their type is used for stack-allocation.
  - They are passed around as an opaque pointer to the stack (or any other caller-allocated space).

As for the types of enums: as you've guessed there are four:
  - An empty array, used for zero-sized enums (zero or one variants, with a ZST payload).
  - A struct with the tag and payload for single-variant enums with non-ZST payloads. I think the tag could be removed here since it should always be `i0` in that case.
  - A struct with the tag and a zero-sized array for C-style enums. As before, I think the payload could be removed since it's always a ZST array.
  - All other cases.

The "all other cases" struct is a bit special since it needs to contain the information about the size and layout without having any actual payload. This is accomplished by making the first field of the struct the smallest integer whose alignment matches the enum's. The remaining bytes are contained within the tailing array. In this case, those fields have nothing to do with the actual data and are only there to allocate the enum correctly within the stack.

All other paddings (ex. between the tag and payload in the representations which contain them) are handled by LLVM.

I'm not sure I understand the enums you've pasted there (the events) so I can't help you with those. If you provide us with a minimal reproducible example (or if it's impossible, the actual failing code, if available of course) of the segfault I'll take a look at it. If possible, I'd prefer a cairo program rather than a contract.

**Comment by DomHenderson** [comment]
Executing main from the following with cairo-native-run produces a segfault, so I think this is a minimal example. I'll work on tidying and pushing the work I've been doing on investigating and fixing this to a branch to avoid duplication of work. I believe the problem to be rooted in build_init and build_match not using the same 4-kind enum representation logic, but that may be a parallel issue
```
enum MyEnum {
    A: felt252,
    B: InnerEnum,
}

enum InnerEnum {
    A: InnerStruct,
}

struct InnerStruct {
    a: felt252,
    b: felt252,
}

fn get_value(e: MyEnum) -> felt252 {
    match e {
        MyEnum::A(a) => a,
        MyEnum::B(a) => match a {
            InnerEnum::A(a) => a.a + a.b
        }
    }
}

fn main() -> (felt252, felt252) {
    (
        get_value(MyEnum::A(10)),
        get_value(MyEnum::B(InnerEnum::A(InnerStruct{a: 30, b: 40}))),
    )
}
```

**Comment by azteca1998** [comment]
I've examined the generated MLIR and LLVM IR from the minimal example and:
  - I've confirmed it segfaults.
  - I've found nothing wrong with it (the generated MLIR is what I expected).
  - Trying to compile it manually with `llc` throws an error.

It seems LLVM doesn't handle an `i0` type correctly, but when used (at least) from the JIT engine it won't throw an error. This seems to be a bug in LLVM. However, since `llc` is failing by not recognizing `i0` as a valid type then we have a bug too. Fixing ours should avoid LLVM's; I'll try to have a fix as soon as possible.

TLDR: There's two bugs, one from LLVM the other on our side. Fixing ours should avoid running into theirs.

**Comment by DomHenderson** [comment]
Oh interesting, thanks for getting on this. In the meantime I'm creating a thorough test suite of every category of enum I can think of just to verify that every case and combination is being handled properly
